tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1839,7 +1839,7 @@ void CGMSHLSLRuntime::AddHLSLFunctionInfo(Function *F, const FunctionDecl *FD) {
       Diags.Report(Attr->getLocation(), DiagID);
       return;
     }
-    if (!isEntry && !isNode) {
+    if ((!SM->IsLib() && !isEntry) && !isNode) {
```

Comment:
I think we need a little more here. If it's a library, we should still verify that it is annotated with `shader("compute")` I'm not sure that's done by the above. The assignment of isCS based on whether its numthreads is pretty suspicious

Reply:
I don't think it's the responsibility of this code in this area to verify that the compute shader attribute exists. For example, I tried changing the shader attribute type to pixel, and later on in the compilation there was an error emitted that there's been an invalid shader attribute combination. So, validation happens later. If you're concerned about there not being validation for non-compute shaders in this specific attribute combination, that validation is covered later on.

Separately, however, I could remove isCS from being set if numthreads is detected, and instead set it if shader("compute") is detected.

Reply:
It's an oddity of how the code is structured that isCS ends up true here since when it finds numthreads and not AS nor MS, it sets isCS to true and only produces an error if there is an entry point. I think isCS shouldn't be true unless that atrrib is there. That could solve a lot of these problems. I'm not sure isNode should determine if this error is displayed at all, but we may have some conversations yet to have over combining those two things

The original bug references "a compute shader entry when compiling to a library target". The `shader("compute")` attrib is what makes it a compute shader entry. As is I think this change will allow WaveSize on any function when compiling to a library. That isn't right either.

Reply:
How much later does the invalid attribute combination message take? We're trying to move a lot of errors sooner, and this is pretty much the last stop of codegen.

Reply:
Line 1989, where const unsigned profileAttributes is defined.

Reply:
This diagnostic shouldn't be generated in CodeGen at all. We should move this into Sema.

Reply:
We should likely land #5467 first then as it provides something of a framework or at least a model by which that could be done

Reply:
Agreed!